ambry-store/src/main/java/com.github.ambry.store/IndexSegment.java

```diff
@@ -662,78 +728,88 @@ private void readFromFile(File fileToRead, Journal journal) throws StoreExceptio
+              + LOG_END_OFFSET_FIELD_LENGTH + CRC_FIELD_LENGTH;
           break;
         default:
           throw new StoreException("IndexSegment : " + indexFile.getAbsolutePath() + " invalid version in index file ",
               StoreErrorCodes.Index_Version_Error);
       }
+      long logEndOffset = stream.readLong();
+      if (version == PersistentIndex.VERSION_0) {
+        lastModifiedTimeSec.set(indexFile.lastModified() / 1000);
+      } else if (version == PersistentIndex.VERSION_1 || version == PersistentIndex.VERSION_2) {
+        lastModifiedTimeSec.set(stream.readLong());
+        StoreKey storeKey = factory.getStoreKey(stream);
+        short resetKeyType = stream.readShort();
+        resetKey = new Pair<>(storeKey, PersistentIndex.IndexEntryType.values()[resetKeyType]);
+        indexSizeExcludingEntries +=
+            (LAST_MODIFIED_TIME_FIELD_LENGTH + resetKey.getFirst().sizeInBytes() + RESET_KEY_TYPE_FIELD_LENGTH);
+      }
+      firstKeyRelativeOffset = indexSizeExcludingEntries - CRC_FIELD_LENGTH;
+      logger.trace("IndexSegment : {} reading log end offset {} from file", indexFile.getAbsolutePath(), logEndOffset);
+      long maxEndOffset = Long.MIN_VALUE;
+      while (stream.available() > CRC_FIELD_LENGTH) {
+        StoreKey key = factory.getStoreKey(stream);
+        byte[] value = new byte[valueSize];
+        stream.read(value);
+        if (version == PersistentIndex.VERSION_2) {
+          byte[] padding = new byte[persistedEntrySize - valueSize - key.sizeInBytes()];
+          stream.read(padding);
+        }
+        IndexValue blobValue = new IndexValue(startOffset.getName(), ByteBuffer.wrap(value), version);
+        long offsetInLogSegment = blobValue.getOffset().getOffset();
+        // ignore entries that have offsets outside the log end offset that this index represents
+        if (offsetInLogSegment + blobValue.getSize() <= logEndOffset) {
+          index.put(key, blobValue);
+          logger.trace("IndexSegment : {} putting key {} in index offset {} size {}", indexFile.getAbsolutePath(), key,
+              blobValue.getOffset(), blobValue.getSize());
+          // regenerate the bloom filter for in memory indexes
+          bloomFilter.add(ByteBuffer.wrap(key.toBytes()));
+          // add to the journal
+          if (blobValue.getOriginalMessageOffset() != IndexValue.UNKNOWN_ORIGINAL_MESSAGE_OFFSET
+              && offsetInLogSegment != blobValue.getOriginalMessageOffset()
+              && blobValue.getOriginalMessageOffset() >= startOffset.getOffset()) {
+            // we add an entry for the original message offset if it is within the same index segment
+            journal.addEntry(new Offset(startOffset.getName(), blobValue.getOriginalMessageOffset()), key);
+          }
+          journal.addEntry(blobValue.getOffset(), key);
+          // sizeWritten is only used for in-memory segments, and for those the padding does not come into picture.
```

Comment:
could this change? Shouldn't we be defining what `sizeWritten` _should_ mean rather than taking current use? Many a time when I have seen this code, I have wondered why the mapped version doesn't support this function.